pkg/kubelet/images/pullmanager/mem_pullrecords.go

```diff
@@ -0,0 +1,186 @@
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"strings"
+	"sync"
+
+	kubeletconfiginternal "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	"k8s.io/utils/lru"
+)
+
+type LRUCache[K comparable, V any] struct {
+	cache *lru.Cache
+}
+
+func NewLRUCache[K comparable, V any](size int) *LRUCache[K, V] {
+	return &LRUCache[K, V]{
+		cache: lru.New(size),
+	}
+}
+
+func (c *LRUCache[K, V]) Get(key K) (*V, bool) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, false
+	}
+	if value == nil {
+		return nil, true
+	}
+	return value.(*V), true
+}
+
+func (c *LRUCache[K, V]) Set(key K, value *V) { c.cache.Add(key, value) }
+func (c *LRUCache[K, V]) Delete(key K)        { c.cache.Remove(key) }
+func (c *LRUCache[K, V]) Len() int            { return c.cache.Len() }
+
+// cachedPullRecordsAccessor implements a write-through cache layer on top
+// of another PullRecordsAccessor
+type cachedPullRecordsAccessor struct {
+	delegate PullRecordsAccessor
+
+	intentsMutex       *sync.RWMutex
+	intents            *LRUCache[string, kubeletconfiginternal.ImagePullIntent]
+	pulledRecordsMutex *sync.RWMutex
+	pulledRecords      *LRUCache[string, kubeletconfiginternal.ImagePulledRecord]
+}
+
+func NewCachedPullRecordsAccessor(delegate PullRecordsAccessor) *cachedPullRecordsAccessor {
+	// TODO: handle errors from records listing here and below
+	coldIntents, _ := delegate.ListImagePullIntents()
+	warmIntents := NewLRUCache[string, kubeletconfiginternal.ImagePullIntent](50)
+	for _, intent := range coldIntents[:min(50, len(coldIntents))] {
+		warmIntents.Set(intent.Image, intent)
+	}
+
+	coldPulledRecords, _ := delegate.ListImagePulledRecords()
+	warmPulledRecords := NewLRUCache[string, kubeletconfiginternal.ImagePulledRecord](100)
+	for _, pulledRecord := range coldPulledRecords[:min(100, len(coldPulledRecords))] {
+		warmPulledRecords.Set(pulledRecord.ImageRef, pulledRecord)
+	}
+
+	return &cachedPullRecordsAccessor{
+		delegate: delegate,
+
+		intentsMutex:       &sync.RWMutex{},
+		intents:            warmIntents,
+		pulledRecordsMutex: &sync.RWMutex{},
+		pulledRecords:      warmPulledRecords,
+	}
+}
+
+func (c *cachedPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
+	//TODO: maybe this should be protected by a read lock
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
+	return c.delegate.ListImagePullIntents()
+}
+
+func (c *cachedPullRecordsAccessor) ImagePullIntentExists(image string) (bool, error) {
+	if _, exists := c.intents.Get(image); exists {
+		return true, nil
+	}
+
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
+
+	exists, err := c.delegate.ImagePullIntentExists(image)
+	if err == nil && exists {
+		c.intents.Set(image, &kubeletconfiginternal.ImagePullIntent{
+			Image: image,
+		})
+	}
+	return exists, err
+}
+
+func (c *cachedPullRecordsAccessor) WriteImagePullIntent(image string) error {
+	c.intentsMutex.Lock()
+	defer c.intentsMutex.Unlock()
```

Comment:
I'm not 100% sure what this lock is for ...
* PullManager already locks its usage so it doesn't concurrently call the recordsAccessor for the same image intent
* fsPullRecordsAccessor is lock-free
* the utils LRU type is thread-safe and locks internally

I think we could possibly drop this, or if this is trying to make it safe to update intents cache from another method, we should maybe use StripedLockSet, so we don't make read/write of independent intents contend for a single intents lock?

Reply:
This was indeed intended to protect the cache when it's updated from the delegate. While it is protected from multiple writes from the PullManager, I wanted to make sure it's also protected as a separately functioning unit.

In case a delete was followed by a write, if context switch occurred after the fs-delete but before mem-delete and wouldn't switch back until the fs+mem write completed, we could possibly end up with a missing cache record.

That may not be an issue in the current implementation, but would be worse should we implement the cache as an authoritative information source.